### PR TITLE
[FLINK-21579][table] Support "SHOW USER FUNCTIONS" statement

### PR DESCRIPTION
**Pull request by yulei0824**
## What is the purpose of the change

This pull request supports listing user defined functions by SHOW USER FUNCTIONS syntax.

## Brief change log

  - Support syntax in both flink and hive parser.
  - Support operation transformation in both old and blink planner.
  - Support syntax in SQL Client.

## Verifying this change

This change added tests and can be verified as follows:
  - Added tests in FlinkSqlParserImplTest and FlinkHiveSqlParserImplTest to verify syntax support.
  - Added tests in SqlToOperatorConverterTest to verify operation transforming behaves correctly, and added tests in TableEnvironmentTest to verify listing user defined function behave corrrectly.
  - Added tests in CliClientTest, SqlCommandParserTest and LocalExecutorITCase to verify SQL Client behaves correctly.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency):  no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper:  no
  - The S3 file system connector:  no

## Documentation

  - Does this pull request introduce a new feature?  yes
  - If yes, how is the feature documented?  docs

### CODE REVIEW

**Comment by wuchong** [comment]
Thanks for the contribution @yulei0824 . However, I think `SHOW USER FUNCTIONS` is a subset of `SHOW FUNCTIONS`. We don't need to implement an independent sytnax. `USER` is just an optional modifier, we may introduce more optional modifier in the future. Could you take https://github.com/apache/flink/pull/15087 this PR as an example ?

**Comment by yulei0824** [comment]
> Thanks for the contribution @yulei0824 . However, I think `SHOW USER FUNCTIONS` is a subset of `SHOW FUNCTIONS`. We don't need to implement an independent sytnax. `USER` is just an optional modifier, we may introduce more optional modifier in the future. Could you take #15087 this PR as an example ?

yes ,  Does support "SHOW [USER] FUNCTIONS" sytanx ?

**Comment by wuchong** [comment]
Yes. That's what I mean.

**Comment by yulei0824** [comment]
Hello @wuchong, Do we implement "SHOW [function_kind] FUNCTIONS" sytanx ? The valid function kinds are USER, SYSTEM and ALL.

**Comment by wuchong** [comment]
@yulei0824 , we only need to support `USER` function_kind in this PR as we discussed in the issue. But the default value of function_kind can be `ALL` (we don't need to support the `ALL` syntax).

**Comment by yulei0824** [comment]
Ok

**Review by wuchong — COMMENTED** [review]
Thanks for the contribution. I left some comments.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ShowFunctionsOperation.java [hunk 1]
```diff
@@ -18,11 +18,21 @@
 
 package org.apache.flink.table.operations;
 
-/** Operation to describe a SHOW FUNCTIONS statement. */
+/** Operation to describe a SHOW [USER] FUNCTIONS statement. */
 public class ShowFunctionsOperation implements ShowOperation {
 
+    private final boolean requireUser;
```
[the remark this answers is not in the document]
Comment (yulei0824) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ShowFunctionsOperation.java:24]: ok

**Comment by wuchong** [comment]
Hi @yulei0824, we just merged an intergration test framework for SQL Client. Now, we don't need to add verbose tests in `CliClientTest` and `LocalExecutorITCase`. Could you revert them and update https://github.com/apache/flink/blob/c4b042f973b3c7a3bbde271eda9620335d8c989f/flink-table/flink-sql-client/src/test/resources/sql/function.q#L23 this file to use `SHOW USER FUNCTIONS`?

**Comment by yulei0824** [comment]
> Hi @yulei0824, we just merged an intergration test framework for SQL Client (#15089). Now, we don't need to add verbose tests in `CliClientTest` and `LocalExecutorITCase`. Could you revert them and update
> 
> https://github.com/apache/flink/blob/c4b042f973b3c7a3bbde271eda9620335d8c989f/flink-table/flink-sql-client/src/test/resources/sql/function.q#L23
> 
> this file to use `SHOW USER FUNCTIONS`?

Yes, I will update the function.q file.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ShowFunctionsOperation.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (yulei0824) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ShowFunctionsOperation.java:24]: What do you think of the following code ?
```
public class ShowFunctionsOperation implements ShowOperation {

    public enum FunctionScope {
        USER,
        ALL
    }

    private final FunctionScope functionScope;
    
    public ShowFunctionsOperation(FunctionScope functionScope) {
        this.functionScope = functionScope;
    }

    @Override
    public String asSummaryString() {
        if (functionScope == FunctionScope.ALL) {
            return "SHOW FUNCTIONS";
        } else {
            return String.format("SHOW %s FUNCTIONS", functionScope);
        }
    }

    public FunctionScope getFunctionScope() {
        return functionScope;
    }
}
```

[the remark this answers is not in the document]
Comment (wuchong) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ShowFunctionsOperation.java:24]: Looks good to me.

[the remark this answers is not in the document]
Comment (yulei0824) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ShowFunctionsOperation.java:24]: Do we add a inner enum FunctionScope to the SqlShowFunctions class ?

[the remark this answers is not in the document]
Comment (wuchong) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ShowFunctionsOperation.java:24]: `SqlShowFunctions` is an internal class. It's fine to update it when we have more scopes. I would prefer to not change it for now.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java [hunk 2]
```diff
@@ -1108,11 +1109,20 @@ private TableResult executeOperation(Operation operation) {
...
+                    showFunctionsOperation.getFunctionScope();
+            switch (functionScope) {
+                case USER:
+                    functionNames = listUserDefinedFunctions();
+                    break;
+                case ALL:
+                    functionNames = listFunctions();
+                    break;
+                default:
+                    throw new UnsupportedFunctionScopeException(functionScope.toString());
```
Comment (wuchong) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java:1123]: We don't need to introduce a new Exception class. Can simply use `new UnsupportedOperationException(String.format("SHOW FUNCTIONS with %s scope is not supported.", scope))` here .

**Review by wuchong — COMMENTED** [review]
It seems that you missed to update `functions.q`.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java (hunk 2 above)
Comment (yulei0824) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java:1123]: Yes

**Comment by yulei0824** [comment]
> It seems that you missed to update `functions.q`.

I will update it on the weekend.

**Comment by yulei0824** [comment]
@flinkbot run azure

**Review by wuchong — APPROVED** [review]
Thanks for the updating. LGTM.

**Comment by wuchong** [comment]
I have rebased and squashed to trigger the build again. Will merge it once the build is passed.

**Comment by yulei0824** [comment]
@wuchong test_ci kafka_gelly failed, This PR does not contain any change about kafka connector.

**Pull request merged**